plugins/autocomplete/plugin.js

```diff
@@ -0,0 +1,1422 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+'use strict';
+
+( function() {
+
+	// This flag prevents appending stylesheet more than once.
+	var stylesLoaded = false;
+
+	CKEDITOR.plugins.add( 'autocomplete', {
+		requires: 'textwatcher',
+		init: function() {
+			if ( !stylesLoaded ) {
+				CKEDITOR.document.appendStyleSheet( this.path + 'skins/' + CKEDITOR.skin.name + '/autocomplete.css' );
+				stylesLoaded = true;
+			}
+		}
+	} );
+
+	/**
+	 * The main class implementing a generic [Autocomplete](https://ckeditor.com/cke4/addon/autocomplete) feature in the editor.
+	 * It is a controller which works with {@link CKEDITOR.plugins.autocomplete.model model} and
+	 * {@link CKEDITOR.plugins.autocomplete.view view} classes.
+	 *
+	 * It's possible to maintain multiple autocomplete instances for a single editor at a time.
+	 * In order to create autocomplete instance use its {@link #constructor constructor}.
+	 *
+	 * @class CKEDITOR.plugins.autocomplete
+	 * @since 4.10.0
+	 * @constructor Creates the new instance of autocomplete and attaches it to the editor.
+	 *
+	 * In order to initialize the autocomplete feature it is enough to instantiate this class with
+	 * two required callbacks:
+	 *
+	 * * A text test callback &ndash; a function which should return a fragment of text (typed in the editor) that
+	 * should be autocompleted (best to use the {@link CKEDITOR.plugins.textMatch#match} function).
+	 * * A data callback &ndash; a function which should return (through its callback) a suggestion data for the current
+	 * query string.
+	 *
+	 * # Creating an autocomplete instance
+	 *
+	 * Depending on your use case, put this code in the {@link CKEDITOR.pluginDefinition#init} callback of your
+	 * plugin or for example in the {@link CKEDITOR.editor#instanceReady} event listener.
+	 *
+	 * ```javascript
+	 *	// Called when the user types in the editor or moves the caret.
```

Comment:
This code is incomplete, as it won't work. Try to provide samples that requires as little work from the reader as possible. Most importantly, sample `itemsArray` should be available.

```js
// Example data feed.
var itemsArray = [ { id: 1, name: '@Andrew' }, { id: 2, name: '@Kate' } ];

// Called when the user types in the editor or moves the caret.
// The range represents the caret position.
function textTestCallback( range ) {
	// We don't want to autocomplete a non-empty selection.
	if ( !range.collapsed ) {
		return null;
	}

	// Use the textmatch plugin which does the tricky job of doing
	// a text search in the DOM. The matchCallback function should return
	// a matching fragment of the text.
	return CKEDITOR.plugins.textMatch.match( range, matchCallback );
}

// Returns a position of the matching text.
// It matches with text starting from the '@' character
// followed by spaces, up to the caret position.
function matchCallback( text, offset ) {
	// Get the text before the caret.
	var left = text.slice( 0, offset ),
		// Will look for an '@' character followed by word characters.
		match = left.match( /@\w*$/ );

	if ( !match ) {
		return null;
	}

	return { start: match.index, end: offset };
}

// Returns (through its callback) the suggestions for the current query.
// Note: the itemsArray variable is our example "database".
function dataCallback( query, range, callback ) {
	// Simple search.
	// Filter the entire items array so only the items that start
	// with the query remain.
	var suggestions = itemsArray.filter( function( item ) {
		return item.name.toLowerCase().indexOf( query.toLowerCase() ) === 0;
	} );

	// Note - the callback function can also be executed asynchronously
	// so dataCallback can do an XHR requests or use any other asynchronous API.
	callback( suggestions );
}

// Finally, instantiate the autocomplete class.
new CKEDITOR.plugins.autocomplete( editor, {
	textTestCallback: textTestCallback,
	dataCallback: dataCallback
} );
```

Reply:
Also it would be nice to mention that user should ensure she or he loads `textmatch` plugin too.